plugin/pkg/shoot/dns/admission.go

```diff
@@ -334,11 +342,35 @@ func assignDefaultDomainIfNeeded(a admission.Attributes, shoot *core.Shoot, proj
 		return apierrors.NewInternalError(err)
 	}
 
-	var shootDomain *string
-	if shoot.Spec.DNS != nil {
-		shootDomain = shoot.Spec.DNS.Domain
+	if shoot.Spec.DNS == nil {
+		shoot.Spec.DNS = &core.DNS{}
+	}
+
+	if len(defaultDomains) > 0 && shoot.Spec.DNS.Domain == nil {
+		domain := defaultDomains[0]
+		shootDNSName := shoot.Name
+
+		if len(shoot.Name) == 0 && len(shoot.GenerateName) > 0 {
+			shootDNSName, err = utils.GenerateRandomStringFromCharset(len(shoot.GenerateName)+5, "0123456789abcdefghijklmnopqrstuvwxyz")
+			if err != nil {
+				return apierrors.NewInternalError(err)
+			}
+		}
+		generatedDomain := fmt.Sprintf("%s.%s.%s", shootDNSName, project.Name, domain)
+		shoot.Spec.DNS.Domain = &generatedDomain
+	}
+
+	return nil
+}
+
+func checkDefaultDomainFormat(a admission.Attributes, shoot *core.Shoot, projectLister corelisters.ProjectLister, defaultDomains []string) error {
+	project, err := admissionutils.ProjectForNamespaceFromInternalLister(projectLister, shoot.Namespace)
+	if err != nil {
+		return apierrors.NewInternalError(err)
 	}
 
+	shootDomain := shoot.Spec.DNS.Domain
+
 	for _, domain := range defaultDomains {
 		if shootDomain != nil && strings.HasSuffix(*shootDomain, "."+domain) {
```

Comment:
You already check whether this is nil before calling the `checkDefaultDomainFormat`. There's no reason to do it again here since the function is not exported.

Reply:
There we are checking shoot.Spec.DNS here it is shoot.Spec.DNS.Domain

Reply:
Ah yes, you are right. In that case can you move this nil check outside of the for loop? I think it's not necessary to be checked on every iteration.
```
shootDomain := shoot.Spec.DNS.Domain
if shootDomian == nil {
    return nil
}
```